targets/TARGET_ARM_SSG/TARGET_MUSCA_A1/serial_api.c

```diff
@@ -0,0 +1,312 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2017-2018 Arm Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <stddef.h>
+/* Use memcpy */
+#include <string.h>
+#include "serial_api.h"
+#include "pinmap.h"
+#include "objects.h"
+/* Use error(...) function */
+#include "mbed_error.h"
+
+#define STDIO_UART_NOT_INITED 0
+#define STDIO_UART_INITED     1
+#define UART_DEVICES_NUMBER   2
+
+struct uart_irq_t {
+    uart_irq_handler handler;
+    uint32_t id;
+};
+
+extern const PinMap PinMap_UART_TX[];
+extern const PinMap PinMap_UART_RX[];
+
+/* Handlers registered */
+static struct uart_irq_t uart_irq[UART_DEVICES_NUMBER];
+
+/* Global variables needed for mbed */
+int stdio_uart_inited = STDIO_UART_NOT_INITED;
+serial_t stdio_uart;
+
+#ifdef UART0_PL011_DEV
+#ifndef uart0_tx_irq_handler
+#error "uart0_tx_irq_handler should be defined, check device_cfg.h!"
+#endif
+void uart0_tx_irq_handler(void)
```

Comment:
are these using static vector placement rather than dynamic (NVIC Set/Get vector functions)? If yes, why?

Reply:
Yes, it's static vector placement. This is a mechanism to make sure the weak function defined in the startup assembly is overriden. So if the expected IRQ handler is not defined, it will be caught by compile time with a proper error message.

Reply:
Correct. I was more after why static one is chosen (we use dynamic one) ?

Reply:
However there is no explicit argument against dynamic vector allocation, this pattern prefers to define platform configuration as early as possible. Since we know the expected behavior in build time, we preferred to set in that step.

It's a commonly used pattern now in our platforms, already upstreamed in CM3DS:
https://github.com/ARMmbed/mbed-os/blob/master/features/netsocket/emac-drivers/TARGET_ARM_SSG/COMPONENT_SMSC9220/smsc9220_emac.cpp#L35
https://github.com/ARMmbed/mbed-os/blob/master/targets/TARGET_ARM_SSG/TARGET_CM3DS_MPS2/us_ticker.c#L111